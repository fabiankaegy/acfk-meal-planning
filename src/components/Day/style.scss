.day {
  background-color: #e5e5e5;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.15);
  border-radius: 15px;
  box-sizing: border-box;
  overflow-y: hidden;
  max-height: 75%;
  min-height: 300px;
  .header {
    background-color: #cde663;
    padding: 15px 20px;
    color: black;
    border-radius: 15px 15px 0px 0px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    h2 {
      padding: 0;
      margin: 0;
    }
    button {
      background: transparent;
      padding: 0;
      border: none;
      height: 20px;
      width: 20px;
      svg {
        height: 100%;
        width: 100%;
        transform: rotate(45deg);
        &:hover {
          filter: brightness(85%);
        }
      }
    }
  }
  .recipes {
    padding: 20px;
    display: flex;
    flex-direction: column;
    > * {
      margin-bottom: 20px;
    }
    max-height: calc(100% - 40px);
    overflow-y: scroll;
  }
}
