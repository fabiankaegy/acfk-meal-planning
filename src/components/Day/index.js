import React, { useContext, useReducer, useEffect } from 'react';
import Popover from '../Popover';
import usePopover from '../Popover/usePopover';
import Button from '../Button';
import Meal from '../Meal';
import { AvailableRecipesContext } from '../AvailableRecipesContext';
import './style.scss';

const recipeReducer = (recipes, action) => {
	switch (action.type) {
		case 'add': {
			return [
				...recipes,
				{
					...action.payload,
					identfier: Date.now(),
				},
			];
		}
		case 'remove': {
			return recipes.filter(recipe => recipe.identfier !== action.payload.identfier);
		}
		case 'clear': {
			return [];
		}
		case 'populate': {
			return [...action.payload];
		}
		default:
			return recipes;
	}
};

const Day = props => {
	const availableRecipes = useContext(AvailableRecipesContext);
	const [recipes, dispatchRecipes] = useReducer(recipeReducer, []);
	const addRecipesPopover = usePopover(false);

	useEffect(() => {
		const recipes = localStorage.getItem(`${props.title}-recipes`);
		if (recipes) {
			dispatchRecipes({
				type: 'populate',
				payload: JSON.parse(recipes),
			});
		}
	}, []);

	useEffect(() => {
		localStorage.setItem(`${props.title}-recipes`, JSON.stringify(recipes));
	}, [recipes]);

	const addRecipe = recipe => {
		dispatchRecipes({ type: 'add', payload: recipe });
	};

	const removeRecipe = identifier => {
		dispatchRecipes({ type: 'remove', payload: identifier });
	};
	return (
		<div className="day">
			<header className="header">
				<h2>{props.title}</h2>
			</header>
			<div className="recipes" data-testid="recipes">
				{recipes.map((recipe, key) => (
					<Meal
						recipe={recipe}
						key={key}
						onClick={() => {
							removeRecipe(recipe);
						}}
					/>
				))}
				<Button onClick={addRecipesPopover.toggle} plus={true} data-testid="add-recipe-button">
					{addRecipesPopover.isShown && (
						<Popover>
							{addRecipesPopover.isShown &&
								availableRecipes &&
								availableRecipes.map((recipe, key) => (
									<Meal onClick={() => addRecipe(recipe)} recipe={recipe} key={key} />
								))}
						</Popover>
					)}
				</Button>
			</div>
		</div>
	);
};

export default Day;
