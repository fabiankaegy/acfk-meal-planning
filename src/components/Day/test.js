import React from 'react';
import ReactDOM from 'react-dom';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import Day from './index';

configure({ adapter: new Adapter() });

describe('Day Component', function() {

    let wrapper;

    beforeEach( function() {
        wrapper = shallow(<Day recipes={[1,2]} title="Title"/>);
    } );

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Day recipes={[1,2]} title="Title"/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })



})
