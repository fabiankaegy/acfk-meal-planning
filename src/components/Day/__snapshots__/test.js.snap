// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day Component matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="day"
      >
        <header
          class="header"
        >
          <h2>
            Title
          </h2>
        </header>
        <div
          class="recipes"
          data-testid="recipes"
        />
        <button
          data-testid="add-recipe-button"
        >
          Add Recipe
        </button>
        <button
          data-testid="remove-recipe-button"
        >
          Remove Recipe
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="day"
    >
      <header
        class="header"
      >
        <h2>
          Title
        </h2>
      </header>
      <div
        class="recipes"
        data-testid="recipes"
      />
      <button
        data-testid="add-recipe-button"
      >
        Add Recipe
      </button>
      <button
        data-testid="remove-recipe-button"
      >
        Remove Recipe
      </button>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
